### PR TITLE
Use IndexExpr to handle dynamic dimensions in lowering SplitOp

### PR DESCRIPTION
**Pull request by tungld**
This patch uses IndexExpr to handle dynamic dimensions in lowering SplitOp.

It also changes ONNXOpShapeHelper to support multiple outputs.  Using `getDimsForOutput(n)` to get and set IndexExpr(s) for the n-th output.

### CODE REVIEW

File: src/Dialect/ONNX/ONNXShapeHelper.hpp [hunk 1]
```diff
@@ -45,12 +47,24 @@ struct ONNXOpShapeHelper {
     llvm_unreachable("implement in child structs");
   }
 
+  // Return output dims for the N-th output.
+  DimsExpr &getDimsForOutput(int n) {
```
Comment (AlexandreEichenberger) [inline src/Dialect/ONNX/ONNXShapeHelper.hpp:51]: Nit: Get/Set goes typically in pairs, and here we have a get that is used for both as you return the reference. So maybe just name it 'dimsForOutput(int n)'. Note that I believe SmallVector already assert for out of bound accesses.

File: test/mlir/onnx/onnx_lowering.mlir [hunk 2]
```diff
@@ -2744,41 +2744,41 @@ func @test_split_unknown_dimension_equal_split(%arg0 : tensor<?x?x64xf32>) -> (t
   %0, %1 = "onnx.Split"(%arg0) { axis = 1 : si64 } : (tensor<?x?x64xf32>) -> (tensor<*xf32>, tensor<*xf32>)
   "std.return"(%0, %1) : (tensor<*xf32>, tensor<*xf32>) -> ()
 
-  // CHECK: [[INDEX_MAP:#.+]] = affine_map<(d0, d1) -> (d0 + d1 ceildiv 2)>
+  // CHECK: [[ALLOC_MAP0:#.+]] = affine_map<(d0, d1) -> (d0 ceildiv 2)>
```
Comment (AlexandreEichenberger) [inline test/mlir/onnx/onnx_lowering.mlir:2747]: It is better to move the test to onnx_lowering_with_canonicalize as the map's math is moved into the code.

Comment (AlexandreEichenberger) [reply test/mlir/onnx/onnx_lowering.mlir:2747]: Also, if you did the FileCheck format by hand, try my new `mlir2FileCheck` utility... it works quite well.

File: src/Dialect/ONNX/ONNXShapeHelper.cpp [hunk 3]
```diff
@@ -364,5 +383,81 @@ LogicalResult ONNXMatMulOpShapeHelper::Compute(
...
+  // Save the final result.
+  getDimsForOutput(0) = outputDims;
+  return success();
+}
+
+//===----------------------------------------------------------------------===//
+// ONNX Split Op Shape Helper
+//===----------------------------------------------------------------------===//
+
+ONNXSplitOpShapeHelper::ONNXSplitOpShapeHelper(
```
Comment (AlexandreEichenberger) [inline src/Dialect/ONNX/ONNXShapeHelper.cpp:395]: Glad it seems to have worked well for your example.

**Review by AlexandreEichenberger — COMMENTED** [review]
Minor fixes, thanks for adding the multiple dim for output.

File: src/Dialect/ONNX/ONNXShapeHelper.hpp (hunk 1 above)
Comment (tungld) [reply src/Dialect/ONNX/ONNXShapeHelper.hpp]: Done.

File: test/mlir/onnx/onnx_lowering.mlir (hunk 2 above)
Comment (tungld) [reply test/mlir/onnx/onnx_lowering.mlir]: `mlir2FileCheck` is quite useful. I used it. Thanks! However, it seemed it did not support local variables that have the same name, e.g. arguments of two krnl loops may use the same name. In the case of SplitOp, I manually changed variable names to make `mlir2FileCheck`work.

**Review by AlexandreEichenberger — APPROVED** [review]
LGTM

**Comment by doru1004** [comment]
This should not have been merged since one of the tests was failing.

**Pull request merged**